src/thermo/BinarySolutionTabulatedThermo.cpp

```diff
@@ -211,27 +258,72 @@ void BinarySolutionTabulatedThermo::initThermoXML(XML_Node& phaseNode, const std
     ThermoPhase::initThermoXML(phaseNode, id_);
 }
 
-std::pair<double, double> BinarySolutionTabulatedThermo::interpolate(double x) const
+double BinarySolutionTabulatedThermo::interpolate(double x,const vector_fp& molefrac,const vector_fp& inputData) const
 {
-    std::pair<double, double> c;
+    double c;
     // Check if x is out of bound
-    if (x > m_molefrac_tab.back()) {
-        c.first = m_enthalpy_tab.back();
-        c.second = m_entropy_tab.back();
+    if (x > molefrac.back()) {
+        c= inputData.back();
         return c;
     }
-    if (x < m_molefrac_tab[0]) {
-        c.first = m_enthalpy_tab[0];
-        c.second = m_entropy_tab[0];
+    if (x < molefrac.front()) {
+        c= inputData.front();
         return c;
     }
-    size_t i = std::distance(m_molefrac_tab.begin(),
-      std::lower_bound(m_molefrac_tab.begin(), m_molefrac_tab.end(), x));
-    c.first = m_enthalpy_tab[i-1] + (m_enthalpy_tab[i] - m_enthalpy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
-    c.second = m_entropy_tab[i-1] + (m_entropy_tab[i] - m_entropy_tab[i-1])
-      * (x - m_molefrac_tab[i-1])/(m_molefrac_tab[i]- m_molefrac_tab[i-1]);
+    size_t i = std::distance(molefrac.begin(),
+      std::lower_bound(molefrac.begin(), molefrac.end(), x));
+    c = inputData[i-1] + (inputData[i] - inputData[i-1])
+      * (x - molefrac[i-1])/(molefrac[i]- molefrac[i-1]);
     return c;
 }
 
+void BinarySolutionTabulatedThermo::integrate(vector_fp& inputData,vector_fp& integratedData,vector_fp& moleFraction)
+{
+	integratedData.resize(inputData.size());
+	for(size_t i = 0; i < integratedData.size(); i++){
+		if (i==0 and moleFraction[0]==0){
+			integratedData[i]=0;
+		}
+		else if (i==0 and moleFraction[0]>0){
+			integratedData[i]=moleFraction[i]*inputData[0];
+		}
+		else{
+			integratedData[i]=integratedData[i-1]+(moleFraction[i]-moleFraction[i-1])/2*(inputData[i]+inputData[i-1]);
+		}
+	}
+}
+
+void BinarySolutionTabulatedThermo::getIntegratedPartialMolarVolumes(doublereal* integrated_vbar)
+{
+	double x[2];
+	IdealSolidSolnPhase::getPartialMolarVolumes(integrated_vbar);
+	Phase::getMoleFractions(x);
+	integrated_vbar[0]=interpolate(x[0],m_molefrac_tab,m_integrated_partial_molar_volume_tab);
+	integrated_vbar[1]=x[1]*integrated_vbar[1];
+}
+
+void BinarySolutionTabulatedThermo::getPartialMolarVolumes(doublereal* vbar) const
+{
+	double x[2];
+	IdealSolidSolnPhase::getPartialMolarVolumes(vbar);
+	Phase::getMoleFractions(x);
+	vbar[0]=interpolate(x[0],m_molefrac_tab,m_partial_molar_volume_tab);
+}
+
+void BinarySolutionTabulatedThermo::calcDensity()
+{
+	// Fallback to IdealSolidSolnPhase::calcDensity() if called before m_integrated_partial_molar_volume_tab is initialized
```

Comment:
I understand why you cannot use the other routine if `m_integrated_partial_molar_volume_tab` is initialized, but wouldn't this still be an erroneous value?  

Is there a specific case where this will be called before the table is initialized, during setup of the class, or is it possible that a user will call it before the initialization is complete?  If the latter, I'd rather find a way for this call to trigger the initialization, or throw an informative error message, rather than returning an incorrect value.

Reply:
That is sort of a workaround. When I implemented the calcDensity() function, the code crashed because the calcDensity() routine got invoked when m_integrated_partial_molar_volume_tab was still empty. 
I wondered how this is possible because actually m_integrated_partial_molar_volume_tab is initialized at the end of void BinarySolutionTabulatedThermo::initThermo(), but I could not track down, which function called calcDensity() before initThermo() initialized the phase. So that was the only solution I could come up with.